src/tests/test_data_service.cpp

```diff
@@ -239,6 +243,162 @@ class BlkDataServiceTest : public testing::Test {
+        shared< Chunk > chunk_in_missing_pdev;
+        shared< Chunk > chunk_in_living_pdev;
+
+        // we can keep all the chunks of a pdev in a vector, but that will be only used for this test,
+        // so, we do not do this for now. we can modify the logic here if we add that vector in the future.
+        for (auto& [_, chunk] : chunks) {
+            if (!chunk) continue;
+            if (chunk_in_missing_pdev && chunk_in_living_pdev) break;
+            if (!chunk_in_missing_pdev && (chunk->physical_dev() == missing_pdev)) {
+                chunk_in_missing_pdev = chunk;
+                continue;
+            }
+            if (!chunk_in_living_pdev && chunk->physical_dev() == living_pdev) {
+                chunk_in_living_pdev = chunk;
+                continue;
+            }
+        }
+
+        RELEASE_ASSERT(chunk_in_missing_pdev, "can not find a chunk on missing drive");
+        RELEASE_ASSERT(chunk_in_living_pdev, "can not find a chunk on living drive");
+
+        LOGINFO("Step 2: write data to these two chunks.");
+        // write blks to both of the chunks
+        MultiBlkId missing_drive_blk;
+        MultiBlkId living_drive_blk;
+
+        blk_alloc_hints hints;
+
+        auto sg_write_ptr1 = std::make_shared< sisl::sg_list >();
+        hints.chunk_id_hint = chunk_in_living_pdev->chunk_id();
+        ++m_outstanding_io_cnt;
+        write_sgs(io_size, sg_write_ptr1, 4, living_drive_blk, hints).thenValue([this](auto&& err) {
+            RELEASE_ASSERT(!err, "Write error");
+            // do not free , use it when test write
+            --m_outstanding_io_cnt;
+            ++m_total_io_comp_cnt;
+        });
+
+        hints.chunk_id_hint = chunk_in_missing_pdev->chunk_id();
+        auto sg_write_ptr2 = std::make_shared< sisl::sg_list >();
+        ++m_outstanding_io_cnt;
+        write_sgs(io_size, sg_write_ptr2, 4, missing_drive_blk, hints).thenValue([this](auto&& err) {
+            RELEASE_ASSERT(!err, "Write error");
+            // free(*sg_write_ptr2); do not free , use it when test write
+            --m_outstanding_io_cnt;
+            ++m_total_io_comp_cnt;
+        });
+
+        // Wait for write operations to complete
+        wait_for_outstanding_io_done();
+
+        LOGINFO("Step 3: restart with missing data drive(pdev).");
+        auto dev_mgr = homestore::HomeStore::instance()->device_mgr();
+        std::vector< std::pair< std::string, homestore::HSDevType > > start_with_devices;
+        auto fast_pdevs = dev_mgr->get_pdevs_by_dev_type(homestore::HSDevType::Fast);
+        for (auto& pdev : fast_pdevs)
+            // can not lose fast drive
+            start_with_devices.emplace_back(pdev->get_devname(), homestore::HSDevType::Fast);
+        // lose one data drive
+        for (auto& pdev : drives) {
+            if (pdev->pdev_id() != missing_pdev->pdev_id())
+                start_with_devices.emplace_back(living_pdev->get_devname(), homestore::HSDevType::Data);
+        }
```

Comment:
i think you need to move the fast/data check in L253 to ensure the missing_pdev you pick is actually a data drive.